### PR TITLE
Make plugin compatible with Config Cache & Test Distribution

### PR DESCRIPTION
**Pull request by paplorinc**
Depends on https://github.com/gradle/test-retry-gradle-plugin/pull/103, fixes https://github.com/gradle/dotcom/issues/9403

### CODE REVIEW

File: plugin/build.gradle.kts [hunk 1]
```diff
@@ -57,12 +58,14 @@ shadowJar.configure {
     }
 }
 
-tasks.getByName("jar").enabled = false
-tasks.getByName("jar").dependsOn(shadowJar)
+tasks.getByName("jar").apply {
```
Comment (bamboo) [inline plugin/build.gradle.kts:61]: 🚀 `tasks.jar { ... }` would be simpler and defer the configuration block

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java [hunk 2]
```diff
@@ -46,9 +48,34 @@ public static void configureTestTask(Test test, ObjectFactory objectFactory, Pro
 
         test.getInputs().property("retry.failOnPassedAfterRetry", adapter.getFailOnPassedAfterRetryInput());
 
+        Provider<Boolean> isDeactivatedByTestDistributionPlugin = shouldTestRetryPluginBeDeactivated(test, objectFactory, providerFactory);
+        test.getInputs().property("isDeactivatedByTestDistributionPlugin", isDeactivatedByTestDistributionPlugin);
```
Comment (bamboo) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:52]: 👍

Comment (bamboo) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:52]: 🦺  Perhaps this should have a corresponding test proving task results are not reused when running with and without test distribution (without changes to the classpath).

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 3]
```diff
@@ -853,7 +840,7 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
         return """
             dependencies {
                 implementation "org.codehaus.groovy:groovy-all:2.5.8"
-                testImplementation "org.spockframework:spock-core:1.3-groovy-2.5"
+                implementation "org.spockframework:spock-core:1.3-groovy-2.5"
```
Comment (bamboo) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:843]: ❓ Why change it from `testImplementation` to `implementation`?

Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:843]: becuse of https://github.com/gradle/test-retry-gradle-plugin/pull/103#discussion_r642856587, but I'll change it to `/test/` instead, I doubt it was deliberate

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java (hunk 2 above)
Comment (paplorinc) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:52]: Good point, I'll do that once #103 is merged

File: plugin/build.gradle.kts (hunk 1 above)
Comment (paplorinc) [reply plugin/build.gradle.kts:61]: Done, thanks

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java (hunk 2 above)
Comment (paplorinc) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:52]: 👍

Comment (paplorinc) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:52]: Added

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java [hunk 4]
```diff
@@ -46,9 +48,53 @@ public static void configureTestTask(Test test, ObjectFactory objectFactory, Pro
...
+
+    public static boolean supportsPropertyConventions(VersionNumber gradleVersion) {
+        return gradleVersion.compareTo(VersionNumber.parse("5.1")) >= 0;
+    }
+
+    private static boolean supportsFinalizeValueOnRead(VersionNumber gradleVersion) {
+        return gradleVersion.compareTo(VersionNumber.parse("6.1")) >= 0;
+    }
+
+    private static boolean callShouldTestRetryPluginBeDeactivated(Test test) {
```
Comment (paplorinc) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:86]: moved from https://github.com/gradle/test-retry-gradle-plugin/pull/104/files#diff-6721c061765073d20229462c6bfb0b45f621ed3fbea5cd173f214869ea18d683L85

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java [hunk 5]
```diff
@@ -46,9 +48,53 @@ public static void configureTestTask(Test test, ObjectFactory objectFactory, Pro
...
+        Test test,
+        ObjectFactory objectFactory,
+        ProviderFactory providerFactory,
+        VersionNumber gradleVersion
+    ) {
+        Provider<Boolean> result = providerFactory.provider(() -> callShouldTestRetryPluginBeDeactivated(test));
+        if (supportsPropertyConventions(gradleVersion)) {
+            Property<Boolean> property = objectFactory.property(Boolean.class).convention(result);
+            if (supportsFinalizeValueOnRead(gradleVersion)) {
+                property.finalizeValueOnRead();
```
Comment (paplorinc) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:71]: memoize

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java [hunk 6]
```diff
@@ -46,9 +48,53 @@ public static void configureTestTask(Test test, ObjectFactory objectFactory, Pro
 
         test.getInputs().property("retry.failOnPassedAfterRetry", adapter.getFailOnPassedAfterRetryInput());
 
+        Provider<Boolean> isDeactivatedByTestDistributionPlugin =
+            shouldTestRetryPluginBeDeactivated(test, objectFactory, providerFactory, gradleVersion);
+        test.getInputs().property("isDeactivatedByTestDistributionPlugin", isDeactivatedByTestDistributionPlugin);
```
Comment (paplorinc) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:53]: does the name matter here?

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java (hunk 5 above)
Comment (snoopcheri) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:71]: ❓ Isn't there way to have a `PropertyWithFinalizedValueOnRead` from the very beginning? I can see how flexible `finalizeValueOnRead()` can be. But at the same time this also looks like you can get it wrong...

File: plugin/src/test/groovy/org/gradle/testretry/ConfigCachingPluginFuncTest.groovy [hunk 7]
```diff
@@ -71,7 +71,39 @@ class ConfigCachingPluginFuncTest extends AbstractGeneralPluginFuncTest {
...
+        configurationCacheIsReused(result, gradleVersion)
+
+        where:
+        gradleVersion << GRADLE_VERSIONS_UNDER_TEST
+    }
+
+    def "compatible with configuration cache when Test Distribution is also present (gradle version #gradleVersion)"() {
+        shouldTestConfigCache(gradleVersion)
+        buildFile << """
+            interface TestDistributionExtension {}
```
Comment (snoopcheri) [inline plugin/src/test/groovy/org/gradle/testretry/ConfigCachingPluginFuncTest.groovy:83]: Cool idea to test it this way!

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java [hunk 8]
```diff
@@ -46,9 +48,53 @@ public static void configureTestTask(Test test, ObjectFactory objectFactory, Pro
...
+            Property<Boolean> property = objectFactory.property(Boolean.class).convention(result);
+            if (supportsFinalizeValueOnRead(gradleVersion)) {
+                property.finalizeValueOnRead();
+            }
+            result = property;
+        }
+        return result;
+    }
+
+    public static boolean supportsPropertyConventions(VersionNumber gradleVersion) {
```
Comment (ldaley) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:78]: After we've merged this, we should centralise these feature checks to a `GradleFeatureAvailability` class or some such.

**Review by ldaley — APPROVED** [review]

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java (hunk 8 above)
Comment (paplorinc) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:78]: Thought about the same, will do it as part of the TD enabling PR

**Review by marcphilipp — APPROVED** [review]
LGTM, thanks for taking care of this! 👍

**Pull request merged**